transaction/DeadLockDetector.cpp

```diff
@@ -0,0 +1,186 @@
+    }
+
+    // TODO(Hakan): Implement logging mechanism for deadlock detection
+    //              start and end times.
+    std::vector<DirectedGraph::node_id> victim_new_batch = getAllVictims();
+
+    // Swap new batch with old batch to make LockTable to see new victims.
+    std::swap(victim_new_batch, victims_);
+
+    // Signal LockTable that new batch is ready.
+    status_.store(DeadLockDetectorStatus::kDone);
+
+    // DeadLockDetector should run once in a predefined interval.
+    std::this_thread::sleep_for(
+        std::chrono::seconds(kSleepDuration));
+  }
+}
+
+void DeadLockDetector::addPendingInfo(transaction_id pending,
+                                      transaction_id owner) {
+  DirectedGraph::node_id pending_node_id = getNodeId(pending);
+  DirectedGraph::node_id owner_node_id = getNodeId(owner);
+
+  // TODO(Hakan): Check first whether link is already created. Use checked
+  //              version for adding an edge.
+  wait_for_graph_->addEdgeUnchecked(pending_node_id, owner_node_id);
+}
+
+void DeadLockDetector::deletePendingInfo(transaction_id pending,
+                                         transaction_id owner) {
+  FATAL_ERROR("Not implemented");
+}
+
+bool DeadLockDetector::isDependent(transaction_id pending, transaction_id owner) {
+  FATAL_ERROR("Not implemented");
+}
+
+std::vector<transaction_id>
+DeadLockDetector::getAllDependents(transaction_id owner) {
+  FATAL_ERROR("Not implemented");
+}
+
+std::vector<transaction_id>
+DeadLockDetector::getAllDependees(transaction_id pending) {
+  DirectedGraph::node_id pending_node_id = getNodeId(pending);
+  std::vector<DirectedGraph::node_id> nodes
+      = wait_for_graph_->getAdjacentNodes(pending_node_id);
+  std::vector<transaction_id> transactions;
+  transactions.reserve(nodes.size());
+  for (DirectedGraph::node_id node_id : nodes) {
+    transaction_id tid = wait_for_graph_->getDataFromNode(node_id);
+    transactions.push_back(tid);
+  }
+  return transactions;
+}
+
+DirectedGraph::node_id DeadLockDetector::getNodeId(transaction_id tid) {
+  DirectedGraph::node_id node_id;
+  if (tid_node_mapping_->count(tid) == 0) {
+    // If it is not created, create it.
+    node_id = addNode(tid);
+  } else {
+    // Otherwise find it in the map.
+    node_id = (*tid_node_mapping_)[tid];
+  }
+  return node_id;
+}
+
+
+DirectedGraph::node_id DeadLockDetector::addNode(transaction_id tid) {
+  transaction_id *tid_ptr = new transaction_id(tid);
+  DirectedGraph::node_id node_id = wait_for_graph_->addNodeUnchecked(tid_ptr);
+  (*tid_node_mapping_)[tid] = node_id;
+  return node_id;
+}
+
+std::vector<transaction_id> DeadLockDetector::getAllVictims() {
+  std::vector<transaction_id> result_victims;
+  wait_for_graph_ = std::make_unique<DirectedGraph>();
+
+  // Critical region on LockTable starts here.
+  lock_table_->latchShared();
+
+  for (LockTable::const_iterator it = lock_table_->begin();
+       it != lock_table_->end(); ++it) {
```

Comment:
No, since we already have `begin` and `end`, the following code works on my Mac with `LLVM`:

```
  for (const auto &lock_item : *lock_table_) {
    const LockTable::lock_own_list &own_list = lock_item.second.first;
    const LockTable::lock_pending_list &pending_list = lock_item.second.second;

    ...
    }
  }
```